packages/env/lib/config.js

```diff
@@ -62,200 +77,281 @@ module.exports = {
+	try {
+		return withBackCompat(
+			JSON.parse( await fs.readFile( configPath, 'utf8' ) )
+		);
+	} catch ( error ) {
+		if ( error.code === 'ENOENT' ) {
+			return null;
+		} else if ( error instanceof SyntaxError ) {
 			throw new ValidationError(
-				'Invalid .wp-env.json: "port" must be an integer.'
+				`Invalid ${ name }: ${ error.message }`
 			);
-		}
-
-		if ( ! Number.isInteger( config.testsPort ) ) {
+		} else {
 			throw new ValidationError(
-				'Invalid .wp-env.json: "testsPort" must be an integer.'
+				`Could not read ${ name }: ${ error.message }`
 			);
 		}
+	}
+}
 
-		if ( config.port === config.testsPort ) {
-			throw new ValidationError(
-				'Invalid .wp-env.json: "testsPort" and "port" must be different.'
-			);
-		}
+/**
+ * Used to maintain back compatibility with older versions of the .wp-env.json
+ * file. Returns an object in the shape of the currently expected .wp-env.json
+ * version.
+ *
+ * @param {Object} rawConfig config right after being read from a file.
+ */
+function withBackCompat( rawConfig ) {
+	// Convert testsPort into new env.tests format.
+	if ( rawConfig.testsPort ) {
+		rawConfig.env = {
+			...( rawConfig.env || {} ),
+			tests: {
+				...( rawConfig.env && rawConfig.env.tests
+					? rawConfig.env.tests
+					: {} ),
+				port: rawConfig.testsPort,
+			},
+		};
+	}
+	return rawConfig;
+}
 
-		if ( typeof config.config !== 'object' ) {
-			throw new ValidationError(
-				'Invalid .wp-env.json: "config" must be an object.'
-			);
-		}
+/**
+ * Overrides keys in the config object with set environment variables or options
+ * which should be merged.
+ *
+ * @param {Config} config fully parsed configuration object.
+ * @return {Config} configuration object with overrides applied.
+ */
+function withOverrides( config ) {
+	// Override port numbers with environment variables.
+	config.env.development.port =
+		getNumberFromEnvVariable( 'WP_ENV_PORT' ) ||
+		config.env.development.port;
+	config.env.tests.port =
+		getNumberFromEnvVariable( 'WP_ENV_TESTS_PORT' ) ||
+		config.env.tests.port;
+
+	// Set the WP_TESTS_DOMAIN port number based on the passed port.
+	const devUrl = new URL( config.env.development.config.WP_TESTS_DOMAIN );
+	devUrl.port = config.env.development.port;
+	config.env.development.config.WP_TESTS_DOMAIN = devUrl.toString();
+
+	const testsUrl = new URL( config.env.tests.config.WP_TESTS_DOMAIN );
+	testsUrl.port = config.env.tests.port;
+	config.env.tests.config.WP_TESTS_DOMAIN = testsUrl.toString();
+
+	return config;
+}
 
-		if ( typeof config.mappings !== 'object' ) {
-			throw new ValidationError(
-				'Invalid .wp-env.json: "mappings" must be an object.'
-			);
-		}
+/**
+ * Parses and validates a config object.
+ *
+ * @param {Object} config A raw config object to parse
+ * @param {string} workDirectoryPath Absolute path to the environment work dir.
+ * @return {Object} the validated and parsed config object.
+ */
+function parseConfig( config, workDirectoryPath ) {
```

Comment:
`parseConfig` now validates a `WPServiceConfig`. this way it can be reused for each environment override. (need to clean up comments around this)